tests/actors/rust/wasi-keyvalue-component/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+            .map_err(|e| format!("failed to parse string from returned bytes: {e}"))?
+            .trim(),
+    )
+    .map_err(|e| format!("failed to convert string to value: {e}"))?;
+
+    Ok(value)
+}
+
+/// Create an outgoing-value to be used by wasi:keyvalue
+fn create_wasi_kv_outgoing_value<T: ToString>(t: T) -> Result<OutgoingValue, String> {
+    let outgoing_value = new_outgoing_value();
+    let stream = outgoing_value_write_body(outgoing_value)
+        .map_err(|_| format!("failed to write outgoing value"))?;
+
+    // Write out the new value
+    write(stream, t.to_string().as_bytes())
+        .map_err(|e| format!("failed to write to outgoing value stream: {e}"))?;
+
+    Ok(outgoing_value)
+}
+
+/// The guest interface for this component defines an actor component that listens for 'Message.Test`
+/// on the wasmCloud lattice.
+///
+/// After receiving that message, the component will run through a suite of tests that validate keyvalue
+/// functionality for the linked wasi:keyvalue provider.
+impl Guest for WasiKeyvalueComponent {
+    fn call(operation: String) -> Result<(), String> {
+        if operation != "WasiKeyValue.Test" {
+            return Err(format!(
+                "invalid operation [{operation}], expecting [WasiKeyValue.Test]"
+            ));
+        }
+
+        // Open a bucket (not supported by some providers, but the call should not fail)
+        let bucket =
+            open_bucket(DEFAULT_BUCKET).map_err(|e| format!("failed to open bucket: {e}"))?;
+        let key = DEFAULT_KEY.into();
+
+        // SET/GET textual values
+        set(
+            bucket,
+            &key,
+            create_wasi_kv_outgoing_value(DEFAULT_TEXT_VALUE)?,
+        )
+        .map_err(|e| format!("failed to set textual value of key [{key}]: {e}"))?;
+        let text: String = read_wasi_kv_incoming_value(get(bucket, &key).map_err(|e| {
+            format!("failed to get textual value of recently set key [{key}]: {e}")
+        })?)?;
+        assert!(
+            text == DEFAULT_TEXT_VALUE,
+            "value retrieved matches the default"
+        );
+
+        // SET/GET/INCREMENT/GET numeric values
+        set(
+            bucket,
+            &key,
+            create_wasi_kv_outgoing_value(DEFAULT_NUMERIC_VALUE)?,
+        )
+        .map_err(|e| format!("failed to set numeric value of key [{key}]: {e}"))?;
+        let number: u64 = read_wasi_kv_incoming_value(get(bucket, &key).map_err(|e| {
+            format!("failed to get numeric value of recently set key [{key}]: {e}")
+        })?)?;
+        assert!(
+            number == 42,
+            "value retrieved matches the incremented value"
+        );
+        let number = increment(bucket, &key, 1)
+            .map_err(|e| format!("failed to get numeric value of recently set key [{key}]: {e}"))?;
+        assert!(
+            number == (DEFAULT_NUMERIC_VALUE + 1).into(),
+            "value retrieved matches the incremented value"
+        );
```

Comment:
This all already exists https://github.com/wasmCloud/wasmCloud/blob/3cef088e82a9c35b2cef76ba34440213361563e4/tests/actors/rust/builtins-component-reactor/src/lib.rs#L186-L260

Reply:
That tests the `builtins`. I saw those tests, and that file is not testing a bindgen'd provider on the lattice -- which is the goal of this PR.

Can you explain how the code path being traversed by by calling `wasmcloud_actor::wasi::keyvalue` is the same as someone would call when building a real bindgen'd provider? This is akin to an integration test where what I am writing is an E2E test.

Reply:
> That tests the `builtins`. I saw those tests, and that file is not testing a bindgen'd provider on the lattice -- which is the goal of this PR.
> 
> Can you explain how the code path being traversed by by calling `wasmcloud_actor::wasi::keyvalue` is the same as someone would call when building a real bindgen'd provider? This is akin to an integration test where what I am writing is an E2E test.

It *has to be* the same, that's the whole point of this effort. Currently, these calls invoke `kvredis` provider built on top of copy-pasted, generated Smithy interfaces. The providers we're using in `test/wasmbus.rs` are by no means less "real" than providers built via wash - we use the exact same functionality to build them.
If you were to update `kvredis` to not depend on copy-pasted interfaces, but rather use interfaces generated from WIT using your bindgen that'd test a real "bindgen'd provider", since that would generate all the structs used in this end-to-end